candidate-agreement/src/table.rs

```diff
@@ -0,0 +1,986 @@
+//! validity and availability votes.
+//!
+//! Each parachain is associated with two sets of validators: those which can
+//! propose and attest to validity of candidates, and those who can only attest
+//! to availability.
+
+use std::collections::HashSet;
+use std::collections::hash_map::{HashMap, Entry};
+use std::hash::Hash;
+use std::fmt::Debug;
+
+/// Context for the statement table.
+pub trait Context {
+	/// A validator ID
+	type ValidatorId: Hash + Eq + Clone + Debug;
+	/// The digest (hash or other unique attribute) of a candidate.
+	type Digest: Hash + Eq + Clone + Debug;
+    /// Candidate type.
+	type Candidate: Ord + Eq + Clone + Debug;
+	/// The group ID type
+	type GroupId: Hash + Ord + Eq + Clone + Debug;
+	/// A signature type.
+	type Signature: Eq + Clone +  Debug;
+
+	/// get the digest of a candidate.
+	fn candidate_digest(&self, candidate: &Self::Candidate) -> Self::Digest;
+
+	/// get the group of a candidate.
+	fn candidate_group(&self, candidate: &Self::Candidate) -> Self::GroupId;
+
+	/// Whether a validator is a member of a group.
+	/// Members are meant to submit candidates and vote on validity.
+	fn is_member_of(&self, validator: &Self::ValidatorId, group: &Self::GroupId) -> bool;
+
+	/// Whether a validator is an availability guarantor of a group.
+	/// Guarantors are meant to vote on availability for candidates submitted
+	/// in a group.
+	fn is_availability_guarantor_of(
+		&self,
+		validator: &Self::ValidatorId,
+		group: &Self::GroupId,
+	) -> bool;
+
+	// requisite number of votes for validity and availability respectively from a group.
+	fn requisite_votes(&self, group: &Self::GroupId) -> (usize, usize);
+}
+
+/// Statements circulated among peers.
+#[derive(PartialEq, Eq, Debug)]
+pub enum Statement<C, D> {
+	/// Broadcast by a validator to indicate that this is his candidate for
+	/// inclusion.
+	///
+	/// Broadcasting two different candidate messages per round is not allowed.
+	Candidate(C),
+	/// Broadcast by a validator to attest that the candidate with given digest
+	/// is valid.
+	Valid(D),
+	/// Broadcast by a validator to attest that the auxiliary data for a candidate
+	/// with given digest is available.
+	Available(D),
+	/// Broadcast by a validator to attest that the candidate with given digest
+	/// is invalid.
+	Invalid(D),
+}
+
+/// A signed statement.
+#[derive(PartialEq, Eq, Debug)]
+pub struct SignedStatement<C, D, V, S> {
+	/// The statement.
+	pub statement: Statement<C, D>,
+	/// The signature.
+	pub signature: S,
+	/// The sender.
+	pub sender: V,
+}
+
+// A unique trace for a class of valid statements issued by a validator.
+//
+// We keep track of which statements we have received or sent to other validators
+// in order to prevent relaying the same data multiple times.
+//
+// The signature of the statement is replaced by the validator because the validator
+// is unique while signatures are not (at least under common schemes like
+// Schnorr or ECDSA).
```

Comment:
(and since we don't need to send the message on, there's no need to keep the much bigger signature over the identity that it is from)

Reply:
the traces don't get sent onwards, but the statements will. we have to keep the signatures for those.